### PR TITLE
Add cdist to pulser.math

### LINKED ISSUE #927: add cdist function to AbstractArray
We currently have the following piece of code in `HamiltonianData`

```
if not positions[0].is_tensor:
            return cast(
                np.ndarray,
                np.round(
                    cast(
                        np.ndarray,
                        cdist(positions, positions, metric="euclidean"),
                    ),
                    COORD_PRECISION,
                ),
            )
        else:
            ten = pm.torch.stack([x._array for x in positions])
            return pm.torch.cdist(ten, ten)
```

here `positions` is a `list[AbstractArray]`. Let's have a function `cdist(list[AbstractArray], list[AbstractArray])` that does the following.

### PR DESCRIPTION
**Pull request by danielgaskins**
## Summary

- Add `pulser.math.cdist()` for NumPy-backed and Torch-backed `AbstractArray` sequences.
- Preserve the Torch computational graph through stacking and distance calculation.
- Use the shared function in `HamiltonianData`.
- Keep the existing NumPy coordinate rounding while leaving Torch distances unrounded so gradients remain available.

## Testing

- `pytest tests/test_math.py -q` (55 passed)
- `pytest tests/test_hamiltonian_data.py -q` (27 passed)
- Black
- isort
- flake8
- mypy
- `git diff --check`

Closes #927

## AI assistance

Full disclosure, I used Codex to inspect the relevant math and Hamiltonian paths, implement the change, draft the tests, and run validation. I reviewed the diff and test results before submitting this pull request.

### CODE REVIEW

File: pulser-core/pulser/_hamiltonian_data/hamiltonian_data.py [hunk 1]
```diff
@@ -172,21 +171,10 @@ def _generate_detuning_fluctuations(
...
-                    cdist(positions, positions, metric="euclidean"),
-                ),
-                COORD_PRECISION,
-            ),
-        )
-    else:
-        ten = pm.torch.stack(
-            [cast(pm.torch.Tensor, x._array) for x in positions]
-        )
-        return pm.AbstractArray(pm.torch.cdist(ten, ten))
```
Comment (a-corni) [inline pulser-core/pulser/_hamiltonian_data/hamiltonian_data.py:189]: @a-quelle-pasqal Is there a reason not to round the distances with torch tensors ?

File: pulser-core/pulser/math/__init__.py [hunk 2]
```diff
@@ -237,6 +237,27 @@ def pdist(a: AbstractArrayLike) -> AbstractArray:
     return AbstractArray(scipy.spatial.distance.pdist(a.as_array()))
 
 
+def cdist(
+    a: Sequence[AbstractArrayLike], b: Sequence[AbstractArrayLike]
+) -> AbstractArray:
+    """Calculate distances between each pair of points in two collections."""
+    abst_a = tuple(map(AbstractArray, a))
+    abst_b = tuple(map(AbstractArray, b))
+    if any(arr.is_tensor for arr in (*abst_a, *abst_b)):
```
Comment (a-corni) [inline pulser-core/pulser/math/__init__.py:246]: Maybe specify in the docstring that torch.cdist is used as soon as one tensor is present in the input sequences ?

**Review by a-corni — COMMENTED** [review]
Hey @danielgaskins ! Nice job! I only have one comment related to documenting your solution, otherwise it looks good to me!

**Comment by danielgaskins** [comment]
Thanks, I added a note that `torch.cdist` is used whenever either input collection contains a torch tensor.
